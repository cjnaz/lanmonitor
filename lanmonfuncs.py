#!/usr/bin/env python3
"""LAN monitor support functions
"""

# __version__ = "V1.0a 210515"

#==========================================================
#
#  Chris Nelson, 2021
#
# V1.1  210523  cmd timeout tweaks
# V1.0  210507  V1.0
# V1.0a 210515  Set timeouts to 1s for ping and 5s for ssh commands on remotes
#
# Changes pending
#   
#==========================================================

import sys
import os.path
import subprocess
import datetime
import time
import re
sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__)), '../funcs3/'))
from funcs3 import logging, getcfg, snd_email, snd_notif

# Configs / Constants
NOTIF_SUBJ = "LAN Monitor"
RTN_PASS     = 0
RTN_WARNING  = 1
RTN_FAIL     = 2
RTN_CRITICAL = 3


# Module global vars
args = None
keylen = 0      # Padding vars used for pretty printing
hostlen = 0


def convert_time (timeX):
    """ Given time value term such as 90s, 15m, 20h, 3d, 2w return int seconds and units.
        <1h> returns (3600, "hours")
    If no units suffix, the value is interpreted as secs.
        <90> returns (90, "secs ")
    """
    if type(timeX) is int:
        return timeX, "secs "           # Case Int
    try:
        return int(timeX), "secs "      # Case Str without units
    except:
        pass
    try:                                # Case Str with units
        time_value = int(timeX[:-1])
        time_unit =  timeX[-1:].lower()
        if time_unit == "s":
            return time_value, "secs "
        if time_unit == "m":
            return time_value * 60, "mins "
        if time_unit == "h":
            return time_value * 60*60, "hours"
        if time_unit == "d":
            return time_value * 60*60*24, "days "
        if time_unit == "w":
            return time_value * 60*60*24*7, "weeks"
        raise ValueError(f"Illegal time units <{time_unit}>")
    except Exception as e:
        logging.error (f"ERROR:  Could not convert time value <{timeX}>\n  {e}")
        raise


USER_HOST_FORMAT = re.compile(r"[\w.-]+@([\w.-]+)$")
USER_HOST_PORT_FORMAT = re.compile(r"[\w.-]+@([\w.-]+):([\d]+)$")

def split_user_host_port(u_h_p):
    """ Handle variations in passed-in user_host_port (u_h_p)
        "local", "user@host", or "user@host:port"
    Return separate user_host, host, and port values
    EG:
        "xyz@host15:4455" returns:
            ["xyz@host15", "host15", "4455"]
        "xyz@host15" returns:
            ["xyz@host15", "host15", "22"]  (the default port is "22" for ssh)
        "local" returns
            ["local", "local", ""]
    """
    user_host_noport = u_h_p.split(":")[0]
    host = u_h_p
    port = ""
    if host != "local":
        port = "22"
        out = USER_HOST_FORMAT.match(u_h_p)
        if out:
            host = out.group(1)
        else:
            out = USER_HOST_PORT_FORMAT.match(u_h_p)
            if out:
                host = out.group(1)
                port = out.group(2)
            else:
                _msg = f"Expecting <user@host> or <user@host:port> format, but found <{u_h_p}>."
                logging.error(f"ERROR:  {_msg}")
                raise ValueError (_msg)
    return user_host_noport, host, port


def cmd_check(cmd, user_host_port, return_type=None, check_line_text=None, expected_text=None, not_text=None):
    """  Runs the cmd and operates on the response based on return_type.
    return_types:
        check_string        Returns True if expected_text occurs in response of the cmd, plus the full full subprocess run structure
            check_line_text     If provided, only the first line containing this text is checked
            expected_text       Text that must be found
            not_text            Text that must NOT be found
        cmdrun              Returns True if the cmd return code was 0, plus the full full subprocess run structure
    """

    if user_host_port != "local":
        u_h, _, port = split_user_host_port(user_host_port)
        cmd = ["ssh", u_h, "-p" + port, "-o", "ConnectTimeout=1", "-T"] + cmd

    for nTry in range (getcfg('nRetries')):
        try:
            logging.debug(f"cmd_check subprocess command try {nTry+1}: <{cmd}>")
            # runtry = subprocess.run(cmd, capture_output=True, text=True)  # Py 3.7+
            runtry = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)   #Py3.6 requires old-style params
        except Exception as e:
            logging.error(f"ERROR:  subprocess.run of cmd <{cmd}> failed.\n  {e}")
            return False, None

        if return_type == "check_string":
            if check_line_text is None:
                text_to_check = runtry.stdout
            else:
                text_to_check = ""
                for line in runtry.stdout.split("\n"):
                    if check_line_text in line:
                        text_to_check = line
                        break
                
            if expected_text in text_to_check:
                if not_text is not None:
                    if not_text not in text_to_check:
                        return True, runtry
                else:
                    return True, runtry
            else:
                if nTry == getcfg('nRetries')-1:
                    return False, runtry

        elif return_type == "cmdrun":
            if runtry.returncode == 0:
                return True, runtry
            elif nTry == getcfg('nRetries')-1:
                return False, runtry

        else:
            _msg = f"Invalid return_type <{return_type}> passed to cmd_check"
            logging.error (f"ERROR:  {_msg}")
            raise ValueError (_msg)
        time.sleep (convert_time(getcfg('RetryInterval'))[0])


have_access_dict = {}

def have_access(user_host_port):
    """If user_host_port != "local" then check ping and ssh access to the host.
    Remember passing result and avoid redoing the ping and ssh tests on later checks.
    If ping/ssh tests fail they will be retried if referenced in a later monitored item.
    The dictionary is cleared on each RecheckInterval in the lanmonitor main code.
    ping or ssh access fails return a tuple of (False, error_message_string).
    (True, "") is returned if local or if ping and ssh tests pass for non-local host.

    NOTE that this optimization can result in confusion:
        if access once passed then fails the plugin will be called and may flag an error
        while the actual problem is no access.
    """
    user_host_port = user_host_port.lower()
    if user_host_port == "local":
        return (True, "")
    if user_host_port in have_access_dict:
        return (True, "")

    _, host, _ = split_user_host_port(user_host_port)
    if host != "local":     # Check for remote access if non-local
        pingrslt = cmd_check(["ping", "-c", "1", "-W", "1", host], user_host_port="local", return_type="cmdrun")
        if not pingrslt[0]:
            return (False, "HOST CANNOT BE REACHED")
        else:
            sshrslt = cmd_check(["echo", "hello"], user_host_port=user_host_port, return_type="cmdrun")
            if "Connection refused" in sshrslt[1].stderr:
                return (False, f"SSH CONNECTION REFUSED <{user_host_port}>")
            # if "The authenticity of host" in sshrslt[1].stdout:       # Never happens thru the script, only interactive
            #     return (False, f"SSH CONNECTION NOT APPROVED.  ssh to <{user_host_port}> manually first.")

    have_access_dict[user_host_port] = True
    return (True, "")


def next_summary_timestring():
    """Calculate timestring of next summary.
    SummaryDays			0 1 2 3 4 5 6	# Days of week.  0 = Sunday
    SummaryTime			9:45		    # 24 hour clock
    NOTE:  May be off by 1 hour over a DTS changes.
    """
    try:
        target_hour   = int(getcfg("SummaryTime","").split(":")[0])
        target_minute = int(getcfg("SummaryTime","").split(":")[1])
        today_day_num = datetime.datetime.today().isoweekday()    # Sunday = 0, Saturday = 6
        now = datetime.datetime.now().replace(second=0, microsecond=0)
        next_summary = now + datetime.timedelta(days=30)
        for daynum in getcfg("SummaryDays").split():
            offset_num_days = int(daynum) - today_day_num
            if offset_num_days < 0:
                offset_num_days += 7
            plus_day =  now + datetime.timedelta(days=offset_num_days)
            with_time = plus_day.replace(hour=target_hour, minute=target_minute)
            if with_time < datetime.datetime.now():
                with_time = with_time + datetime.timedelta(days=7)
            if with_time < next_summary:
                next_summary = with_time
        logging.debug(f"Next summary:  {next_summary}")
        return next_summary
    except Exception as e:
        _msg = f"SummaryDays <{getcfg('SummaryDays','')}> or SummaryTime <{getcfg('SummaryTime','')}> settings could not be parsed\n  {e}"
        logging.error(f"ERROR:  {_msg}")
        raise ValueError (_msg) from None
        