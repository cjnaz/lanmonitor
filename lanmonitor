#!/usr/bin/env python3
"""LAN monitor

Monitor these local network items, and send notification when something doesn't look right:
    SELinux status
    Hosts ping response
    Systemd services active and running
    Web pages responding with expected text
    Processes existing
    Filesystem age

Operates as a systemd service, or interactively with --once switch.
"""

__version__ = "V0.4 210304"

#==========================================================
#
#  Chris Nelson, 2021
#
# V0.4 200304  Added StartupDelay for service mode.  Added SELinux check.
# V0.3 200226  Bug fix for files mod time vs. create time
# V0.2 210207  Added age info to Activity log
# V0.1 210129  Initial
#
# Changes pending
#   
#==========================================================

import argparse
import sys
import subprocess
import glob
import requests
import os.path
import signal       # For keyboard interrupt handler

sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__)), '../funcs3/'))
from funcs3 import *


# Configs / Constants
PY_MIN_VERSION = 3.6
FUNCS3_MIN_VERSION = 0.5
NOTIF_SUBJ = "LAN Monitor"
CONFIG_FILE = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'lanmonitor.cfg')
LOG_FILE = "log_lanmonitor.txt"

py_version = float(sys.version_info.major) + float(sys.version_info.minor)/10
if py_version < PY_MIN_VERSION:
    print (f"ERROR:  Current Python version {py_version} is less than minimum required version {PY_MIN_VERSION} - Aborting")
    sys.exit(1)


def main():
    notifs_sent = {}
    first = True

    while 1:
        loadconfig(cfgfile = CONFIG_FILE) #, cfgloglevel=10)

        if args.once:
            logging.getLogger().setLevel(logging.INFO)
        else:
            if first:
                time.sleep (getcfg('StartupDelay'))
                first = False


        def handle_item(good, key, good_message, bad_message):
            """
            good:  Boolean, where True means that the status is good
            key:   Key into notifs_sent dictionary which stores notification times
            """
            if not good:
                do_send = False
                if key not in notifs_sent:
                    do_send = True
                else:
                    if time.time() > notifs_sent[key]:
                        do_send = True
                if do_send and not args.once:
                    snd_notif (subj=NOTIF_SUBJ, msg=bad_message, log=True)
                    notifs_sent[key] = time.time() + getcfg("ReNotificationInterval")
                else:
                    logging.info(bad_message)
            else:
                logging.info(good_message)


        # Check SELinux state
        if "SELinux" in cfg:
            handle_item (check_selinux(), "SELINUX",
                good_message = f"SELinux OK:  {getcfg('SELinux')}",
                bad_message  = f"SELinux is NOT in expected state ({getcfg('SELinux')})" )

        # Check Hosts
        for host in cfg:
            if host.startswith("Host_"):
                host_tag = host[5:]
                ip = cfg[host]
                handle_item(ping_host(ip), "HOST_" + host,
                    good_message = f"Host OK:  {host_tag:14} at {ip}",
                    bad_message  = f"Host system <{host_tag}> at {ip} is NOT responding" )

        # Check Services
        for service in getcfg("Services", "").split():
            handle_item(check_service(service), "SERVICE_" + service,
                good_message = f"Service OK:  {service}",
                bad_message  = f"Service <{service}> is NOT running")

        # Check Pages
        for page in cfg:
            if page.startswith("Page_"):
                page_tag = page[5:]
                url = cfg[page].split()[0]
                string = cfg[page][len(url):].strip().rstrip()
                handle_item (check_page(url, string), "PAGE_" + page_tag,
                    good_message = f"Page OK:  {page_tag}",
                    bad_message  = f"Page <{page_tag}> is NOT running" )

        # Check Processes
        for process in cfg:
            if process.startswith("Process_"):
                process_tag = process[8:]
                full_path = cfg[process]
                handle_item (check_process(process_tag, full_path), "PROCESS_" + process_tag,
                    good_message = f"Process OK:  {process_tag}",
                    bad_message  = f"Process <{process_tag}> is NOT running" )

        # Check for stale directories/files
        now = time.time()
        for dirpath in cfg:
            if dirpath.startswith("Activity_"):
                dirpath_tag = dirpath[9:]
                max_age_days = cfg[dirpath].split()[0]
                the_path = cfg[dirpath][len(cfg[dirpath].split()[0]):].strip().rstrip()
                if os.path.isdir(the_path):
                    if the_path.endswith("/"):
                        the_path += "*"
                    else:
                        the_path += "/*"
                elif not os.path.isfile(the_path):
                    logging.warning(f"WARNING:  Activity tag <{dirpath_tag}> ({the_path}) is not valid or not accessible.  Skipping")
                    continue

                latest_file_modtime = os.path.getmtime(max(glob.glob(the_path), key=os.path.getmtime))
                age = (now - latest_file_modtime) /60/60/24
                handle_item (latest_file_modtime > now - 60*60*24*int(max_age_days), "ACTIVITY_" + dirpath_tag,
                    good_message = f"File activity OK:         {dirpath_tag:18}  {age:4.1f} days, max {max_age_days:>2} days  {the_path}",
                    bad_message  = f"STALE FILES at Activity:  {dirpath_tag:18}  {age:4.1f} days, max {max_age_days:>2} days  {the_path}")

        if args.once:
            sys.exit(0)

        time.sleep (getcfg('RecheckInterval'))
    


def check_selinux():
    """Check that sestatus returns Current mode matching value from cfg.
    Return True if matching.
    """
    for _ in range (getcfg('nRetries')):
        try:
            _cmd = ["sestatus"]
            ps_response = subprocess.run(_cmd, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True).stdout.split("\n")
            for line in ps_response:
                if "Current mode:" in line:
                    if getcfg("SELinux") in line:
                        return True
        except Exception as e:
            logging.info ("Exception:\n  " + repr(e))
    return False


def ping_host(host_ip):
    """Ping the supplied IP address.
    Return True if ping successful, else False.
    """
    for _ in range (getcfg('nRetries')):
        try:
            _cmd = ["ping", host_ip, "-c", "1"]
            host_up  = True if subprocess.run(_cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL).returncode is 0 else False
            if host_up:
                return True
        except Exception as e:
            logging.info ("Exception:\n  " + repr(e))
    return False


def check_service(service_name):
    """Check the active status of a service.
    Return True if "active (running)"
    """
    for _ in range (getcfg('nRetries')):
        try:
            _cmd = ["systemctl", "status", service_name]
            status = subprocess.run(_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True).stdout
            if "active (running)" in status:
                return True
            else:
                return False
        except Exception as e:
            logging.info ("Exception:\n  " + repr(e))
    return False


def check_page(url, text):
    """Check web page containing text.
    Return True if page contains text.
    """
    for _ in range (getcfg('nRetries')):
        try:
            web_page = requests.get(url, timeout=0.5).text
            return text in web_page
        except requests.exceptions.ConnectionError as e:
            msg = "ConnectionError\n  " + repr(e)
        except requests.exceptions.Timeout as e:
            msg = "Timeout\n  " + repr(e)
        except Exception as e:
            msg = "Other exception\n  " + repr(e)
    logging.info (msg)
    return False


def check_process(process_name, path):
    """Check for running process.
    Return True if ps response has a line containing path.
    """
    for _ in range (getcfg('nRetries')):
        try:
            _cmd = ["ps", "-Af"]
            ps_response = subprocess.run(_cmd, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True).stdout.split("\n")
            for line in ps_response:
                if path in line:
                    return True
        except Exception as e:
            logging.info ("Exception:\n  " + repr(e))
    return False


def keyboardInterruptHandler(signal, frame):
    print("KeyboardInterrupt (ID: {}) has been caught. Cleaning up...".format(signal))
    sys.exit(0)
signal.signal(signal.SIGINT, keyboardInterruptHandler)
    

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description=__doc__ + __version__, formatter_class=argparse.RawTextHelpFormatter)
    parser.add_argument('-1', '--once', action='store_true',
                        help="Single run mode.  Logging is to console rather than file.")
    parser.add_argument('-V', '--version', action='version', version='%(prog)s ' + __version__,
                        help="Return version number and exit.")

    args = parser.parse_args()

    if args.once:
        setuplogging(logfile=None)
    else:
        setuplogging(logfile=LOG_FILE)

    if not funcs3_min_version_check(FUNCS3_MIN_VERSION):
        logging.error(f"ERROR:  funcs3 module must be at least version {FUNCS3_MIN_VERSION}.  Found <{funcs3_version}>.")
        sys.exit(1)

    main()
    sys.exit()