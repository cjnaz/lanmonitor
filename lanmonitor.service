[Unit]
Description=Monitor processes and devices on the LAN
After=network-online.target

[Service]
ExecStart=/<path to>/lanmonitor

[Install]
WantedBy=multi-user.target
