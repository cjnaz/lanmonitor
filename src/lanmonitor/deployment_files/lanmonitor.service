[Unit]
Description=Monitor processes and devices on the LAN
After=network-online.target

[Service]
ExecStart=/<path_to>/lanmonitor --service

[Install]
WantedBy=multi-user.target
