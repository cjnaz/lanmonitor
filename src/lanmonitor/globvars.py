#!/usr/bin/env python3
"""LAN monitor global vars
"""

#==========================================================
#
#  Chris Nelson, Copyright 2021-2023
#
# 3.0 230301 - Packaged
#   
#==========================================================

args = None
keylen = 0
hostlen = 0
