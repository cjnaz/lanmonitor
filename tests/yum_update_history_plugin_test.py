#!/usr/bin/env python3
"""LAN Monitor plugin - yum_update_history_plugin_test
"""

#==========================================================
#
#  Chris Nelson, Copyright 2021-2024
#
# 3.3 240805 - Updated to lanmonitor V3.3
# 3.1 230320 - Added ssh access warning cases
# 3.0 230301 - Packaged
#
#==========================================================

from cjnfuncs.core import set_toolname, setuplogging, logging
from cjnfuncs.configman import config_item
import lanmonitor.globvars as globvars
from lanmonitor.lanmonfuncs import RTN_PASS
try:
    from yum_update_history_plugin import monitor
except:
    from lanmonitor.yum_update_history_plugin import monitor

set_toolname('tool')
globvars.config = config_item()
globvars.config.cfg['nTries']           = 1
globvars.config.cfg['RetryInterval']    = '0s'
globvars.config.cfg['SSH_timeout']      = '4s'
setuplogging(ConsoleLogFormat="{module:>35}.{funcName:20} - {levelname:>8}:  {message}")
logging.getLogger().setLevel(logging.DEBUG)

def dotest (tnum, desc, test):
    print(f"\nTest {tnum} - {desc} {'-'*50}")
    inst = monitor()
    setup_rslt = inst.setup(test)
    logging.debug (f"{test['key']} - setup() returned:  {setup_rslt}")
    if setup_rslt == RTN_PASS:
        logging.debug (f"{test['key']} - eval_status() returned:  {inst.eval_status()}")

dotest (1, "YumUpdate pass - OK",
        {'key':'YumUpdate_Pass', 'tag':'Pass', 'host':'local', 'user_host_port':'local', 'critical':True, 'cmd_timeout':2, 'check_interval':1, 'rest_of_line':'   90d update  '})

dotest (2, "YumUpdate too old - CRITICAL",
        {'key':'YumUpdate_TooOld', 'tag':'TooOld', 'host':'local', 'user_host_port':'local', 'critical':True, 'cmd_timeout':2, 'check_interval':1, 'rest_of_line':'10m update'})

dotest (3, "Bad definition - setup ERROR",
        {'key':'YumUpdate_baddef', 'tag':'badline', 'host':'local', 'user_host_port':'local', 'critical':True, 'cmd_timeout':2, 'check_interval':1, 'rest_of_line':'10m'})

dotest (4, "Bad time - setup ERROR",
        {'key':'YumUpdate_badtime', 'tag':'badtime', 'host':'local', 'user_host_port':'local', 'critical':True, 'cmd_timeout':2, 'check_interval':1, 'rest_of_line':'10y update'})

dotest (5, "Remote not running yum - Can't get history - WARNING",
        {'key':'YumUpdate_remote', 'tag':'remote', 'host':'testhost2', 'user_host_port':'me@testhost2', 'critical':True, 'cmd_timeout':2, 'check_interval':1, 'rest_of_line':'10m update'})

dotest (6, "No matching command history - WARNING",
        {'key':'YumUpdate_noupdates', 'tag':'noupdates', 'host':'local', 'user_host_port':'local', 'critical':True, 'cmd_timeout':2, 'check_interval':1, 'rest_of_line':'10m update xx'})

dotest (7, "No such host - WARNING",
        {'key':'YumUpdate_Unknown', 'tag':'Unknown', 'host':'nosuchhost', 'user_host_port':'me@nosuchhost', 'critical':True, 'cmd_timeout':4, 'check_interval':1, 'rest_of_line':'90d update'})

dotest (8, "Known host, unavailable - WARNING",
        {'key':'YumUpdate_Unavailable', 'tag':'Unavailable', 'host':'testhostX', 'user_host_port':'me@testhostX', 'critical':True, 'cmd_timeout':4, 'check_interval':1, 'rest_of_line':'90d update'})
