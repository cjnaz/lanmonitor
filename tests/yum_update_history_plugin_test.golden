
Test 1 - YumUpdate pass - OK --------------------------------------------------
          yum_update_history_plugin.setup                -    DEBUG:  YumUpdate_Pass - lanmonitor.yum_update_history_plugin.setup() called:
  {'key': 'YumUpdate_Pass', 'tag': 'Pass', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '   90d update  '}
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_Pass - setup() returned:  0
          yum_update_history_plugin.eval_status          -    DEBUG:  YumUpdate_Pass - lanmonitor.yum_update_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['yum', 'history']>
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_Pass - eval_status() returned:  {'rslt': 0, 'notif_key': 'YumUpdate_Pass', 'message': 'YumUpdate_Pass  OK - local -   36.2 days  (  90 days  max)'}

Test 2 - YumUpdate too old - CRITICAL --------------------------------------------------
          yum_update_history_plugin.setup                -    DEBUG:  YumUpdate_TooOld - lanmonitor.yum_update_history_plugin.setup() called:
  {'key': 'YumUpdate_TooOld', 'tag': 'TooOld', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '10m update'}
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_TooOld - setup() returned:  0
          yum_update_history_plugin.eval_status          -    DEBUG:  YumUpdate_TooOld - lanmonitor.yum_update_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['yum', 'history']>
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_TooOld - eval_status() returned:  {'rslt': 3, 'notif_key': 'YumUpdate_TooOld', 'message': '  CRITICAL: YumUpdate_TooOld - local - YUM UPDATE TOO LONG AGO - 52171.5 mins  (  10 mins  max)'}

Test 3 - Bad definition - setup ERROR --------------------------------------------------
          yum_update_history_plugin.setup                -    DEBUG:  YumUpdate_baddef - lanmonitor.yum_update_history_plugin.setup() called:
  {'key': 'YumUpdate_baddef', 'tag': 'badline', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '10m'}
          yum_update_history_plugin.setup                -    ERROR:    ERROR:  <YumUpdate_baddef> INVALID LINE SYNTAX <10m>
  list index out of range
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_baddef - setup() returned:  2

Test 4 - Bad time - setup ERROR --------------------------------------------------
          yum_update_history_plugin.setup                -    DEBUG:  YumUpdate_badtime - lanmonitor.yum_update_history_plugin.setup() called:
  {'key': 'YumUpdate_badtime', 'tag': 'badtime', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '10y update'}
          yum_update_history_plugin.setup                -    ERROR:    ERROR:  <YumUpdate_badtime> INVALID LINE SYNTAX <10y update>
  Illegal time units <y> in time string <10y>
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_badtime - setup() returned:  2

Test 5 - Remote not running yum - Can't get history - WARNING --------------------------------------------------
          yum_update_history_plugin.setup                -    DEBUG:  YumUpdate_remote - lanmonitor.yum_update_history_plugin.setup() called:
  {'key': 'YumUpdate_remote', 'tag': 'remote', 'host': 'testhost2', 'user_host_port': 'me@testhost2', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '10m update'}
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_remote - setup() returned:  0
          yum_update_history_plugin.eval_status          -    DEBUG:  YumUpdate_remote - lanmonitor.yum_update_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@testhost2', '-p22', '-T', 'yum', 'history']>
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 failed (returncode=127) <['ssh', 'me@testhost2', '-p22', '-T', 'yum', 'history']>:  bash: line 1: yum: command not found
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd failed on remote system with <bash: line 1: yum: command not found> - attempting simplessh connection to <me@testhost2>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 <['ssh', 'me@testhost2', '-p22', '-T', 'echo', 'hello']>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh connection passes.  Return original cmd failure record.
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_remote - eval_status() returned:  {'rslt': 1, 'notif_key': 'YumUpdate_remote', 'message': '  WARNING: YumUpdate_remote - testhost2 - COULD NOT GET YUM HISTORY'}

Test 6 - No matching command history - WARNING --------------------------------------------------
          yum_update_history_plugin.setup                -    DEBUG:  YumUpdate_noupdates - lanmonitor.yum_update_history_plugin.setup() called:
  {'key': 'YumUpdate_noupdates', 'tag': 'noupdates', 'host': 'local', 'user_host_port': 'local', 'critical': True, 'cmd_timeout': 2, 'check_interval': 1, 'rest_of_line': '10m update xx'}
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_noupdates - setup() returned:  0
          yum_update_history_plugin.eval_status          -    DEBUG:  YumUpdate_noupdates - lanmonitor.yum_update_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['yum', 'history']>
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_noupdates - eval_status() returned:  {'rslt': 1, 'notif_key': 'YumUpdate_noupdates', 'message': '  WARNING: YumUpdate_noupdates - local - NO <update xx> UPDATES IN THE YUM HISTORY DB'}

Test 7 - No such host - WARNING --------------------------------------------------
          yum_update_history_plugin.setup                -    DEBUG:  YumUpdate_Unknown - lanmonitor.yum_update_history_plugin.setup() called:
  {'key': 'YumUpdate_Unknown', 'tag': 'Unknown', 'host': 'nosuchhost', 'user_host_port': 'me@nosuchhost', 'critical': True, 'cmd_timeout': 4, 'check_interval': 1, 'rest_of_line': '90d update'}
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_Unknown - setup() returned:  0
          yum_update_history_plugin.eval_status          -    DEBUG:  YumUpdate_Unknown - lanmonitor.yum_update_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@nosuchhost', '-p22', '-T', 'yum', 'history']>
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 failed (returncode=255) <['ssh', 'me@nosuchhost', '-p22', '-T', 'yum', 'history']>:  ssh: Could not resolve hostname nosuchhost: Name or service not known
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd failed on remote system with <ssh: Could not resolve hostname nosuchhost: Name or service not known> - attempting simplessh connection to <me@nosuchhost>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 <['ssh', 'me@nosuchhost', '-p22', '-T', 'echo', 'hello']>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 failed (returncode=255) <['ssh', 'me@nosuchhost', '-p22', '-T', 'echo', 'hello']>:  ssh: Could not resolve hostname nosuchhost: Name or service not known
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh failed with <ssh: Could not resolve hostname nosuchhost: Name or service not known>
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_Unknown - eval_status() returned:  {'rslt': 1, 'notif_key': 'YumUpdate_Unknown', 'message': '  WARNING: YumUpdate_Unknown - nosuchhost - ssh: Could not resolve hostname nosuchhost: Name or service not known'}

Test 8 - Known host, unavailable - WARNING --------------------------------------------------
          yum_update_history_plugin.setup                -    DEBUG:  YumUpdate_Unavailable - lanmonitor.yum_update_history_plugin.setup() called:
  {'key': 'YumUpdate_Unavailable', 'tag': 'Unavailable', 'host': 'testhostX', 'user_host_port': 'me@testhostX', 'critical': True, 'cmd_timeout': 4, 'check_interval': 1, 'rest_of_line': '90d update'}
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_Unavailable - setup() returned:  0
          yum_update_history_plugin.eval_status          -    DEBUG:  YumUpdate_Unavailable - lanmonitor.yum_update_history_plugin.eval_status() called
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 <['ssh', 'me@testhostX', '-p22', '-T', 'yum', 'history']>
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd try 1 timeout:  Command '['ssh', 'me@testhostX', '-p22', '-T', 'yum', 'history']' timed out after 4 seconds
                        lanmonfuncs.cmd_check            -    DEBUG:  cmd failed on remote system with <Command '['ssh', 'me@testhostX', '-p22', '-T', 'yum', 'history']' timed out after 4 seconds> - attempting simplessh connection to <me@testhostX>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 <['ssh', 'me@testhostX', '-p22', '-T', 'echo', 'hello']>
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh try 1 timeout:  Command '['ssh', 'me@testhostX', '-p22', '-T', 'echo', 'hello']' timed out after 4.0 seconds
                        lanmonfuncs.cmd_check            -    DEBUG:  simplessh failed with <Command '['ssh', 'me@testhostX', '-p22', '-T', 'echo', 'hello']' timed out after 4.0 seconds>
     yum_update_history_plugin_test.dotest               -    DEBUG:  YumUpdate_Unavailable - eval_status() returned:  {'rslt': 1, 'notif_key': 'YumUpdate_Unavailable', 'message': "  WARNING: YumUpdate_Unavailable - testhostX - cmd_check simplessh timeout:  Command '['ssh', 'me@testhostX', '-p22', '-T', 'echo', 'hello']' timed out after 4.0 seconds"}
